File: plugins/markdown/core/src/org/intellij/plugins/markdown/ui/split/SplitTextEditorProvider.java
// Copyright 2000-2018 JetBrains s.r.o.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package org.intellij.plugins.markdown.ui.split;

import com.intellij.openapi.fileEditor.*;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class SplitTextEditorProvider implements AsyncFileEditorProvider, DumbAware {

  protected static final String FIRST_EDITOR = "first_editor";
  protected static final String SECOND_EDITOR = "second_editor";
  protected static final String SPLIT_LAYOUT = "split_layout";

  @NotNull
  protected final FileEditorProvider myFirstProvider;
  @NotNull
  protected final FileEditorProvider mySecondProvider;

  @NotNull
  private final String myEditorTypeId;

  public SplitTextEditorProvider(@NotNull FileEditorProvider firstProvider, @NotNull FileEditorProvider secondProvider) {
    myFirstProvider = firstProvider;
    mySecondProvider = secondProvider;

    myEditorTypeId = "split-provider[" + myFirstProvider.getEditorTypeId() + ";" + mySecondProvider.getEditorTypeId() + "]";
  }

  @Override
  public boolean accept(@NotNull Project project, @NotNull VirtualFile file) {
    return myFirstProvider.accept(project, file) && mySecondProvider.accept(project, file);
  }

  @NotNull
  @Override
  public FileEditor createEditor(@NotNull Project project, @NotNull VirtualFile file) {
    return createEditorAsync(project, file).build();
  }

  @NotNull
  @Override
  public String getEditorTypeId() {
    return myEditorTypeId;
  }

  @NotNull
  @Override
  public Builder createEditorAsync(@NotNull final Project project, @NotNull final VirtualFile file) {
    final Builder firstBuilder = getBuilderFromEditorProvider(myFirstProvider, project, file);
    final Builder secondBuilder = getBuilderFromEditorProvider(mySecondProvider, project, file);

    return new Builder() {
      @Override
      public FileEditor build() {
        return createSplitEditor(firstBuilder.build(), secondBuilder.build());
      }
    };
  }

  @Nullable
  protected FileEditorState readFirstProviderState(@NotNull Element sourceElement, @NotNull Project project, @NotNull VirtualFile file) {
    final var child = sourceElement.getChild(FIRST_EDITOR);
    if (child != null) {
      return myFirstProvider.readState(child, project, file);
    }
    return null;
  }

  @Nullable
  protected FileEditorState readSecondProviderState(@NotNull Element sourceElement, @NotNull Project project, @NotNull VirtualFile file) {
    final var child = sourceElement.getChild(SECOND_EDITOR);
    if (child != null) {
      return mySecondProvider.readState(child, project, file);
    }
    return null;
  }

  @Nullable
  protected String readSplitLayoutState(@NotNull Element sourceElement, @NotNull Project project, @NotNull VirtualFile file) {
    final var attribute = sourceElement.getAttribute(SPLIT_LAYOUT);
    String layoutName = null;
    if (attribute != null) {
      layoutName = attribute.getValue();
    }
    return layoutName;
  }

  @NotNull
  @Override
  public FileEditorState readState(@NotNull Element sourceElement, @NotNull Project project, @NotNull VirtualFile file) {
    final var firstState = readFirstProviderState(sourceElement, project, file);
    final var secondState = readSecondProviderState(sourceElement, project, file);
    final var layoutName = readSplitLayoutState(sourceElement, project, file);
    return new SplitFileEditor.MyFileEditorState(layoutName, firstState, secondState);
  }

  protected void writeFirstProviderState(@Nullable FileEditorState state, @NotNull Project project, @NotNull Element targetElement) {
    final var child = new Element(FIRST_EDITOR);
    if (state != null) {
      myFirstProvider.writeState(state, project, child);
      targetElement.addContent(child);
    }
  }

  protected void writeSecondProviderState(@Nullable FileEditorState state, @NotNull Project project, @NotNull Element targetElement) {
    final var child = new Element(SECOND_EDITOR);
    if (state != null) {
      mySecondProvider.writeState(state, project, child);
      targetElement.addContent(child);
    }
  }

  protected void writeSplitLayoutState(@Nullable String splitLayout, @NotNull Project project, @NotNull Element targetElement) {
    if (splitLayout != null) {
      targetElement.setAttribute(SPLIT_LAYOUT, splitLayout);
    }
  }

  @Override
  public void writeState(@NotNull FileEditorState state, @NotNull Project project, @NotNull Element targetElement) {
    if (!(state instanceof SplitFileEditor.MyFileEditorState)) {
      return;
    }
    final var compositeState = (SplitFileEditor.MyFileEditorState)state;
    writeFirstProviderState(compositeState.getFirstState(), project, targetElement);
    writeSecondProviderState(compositeState.getSecondState(), project, targetElement);
    writeSplitLayoutState(compositeState.getSplitLayout(), project, targetElement);
  }

  protected abstract FileEditor createSplitEditor(@NotNull FileEditor firstEditor, @NotNull FileEditor secondEditor);

  @NotNull
  @Override
  public FileEditorPolicy getPolicy() {
    return FileEditorPolicy.HIDE_DEFAULT_EDITOR;
  }

  @NotNull
  public static Builder getBuilderFromEditorProvider(@NotNull final FileEditorProvider provider,
                                                     @NotNull final Project project,
                                                     @NotNull final VirtualFile file) {
    if (provider instanceof AsyncFileEditorProvider) {
      return ((AsyncFileEditorProvider)provider).createEditorAsync(project, file);
    }
    else {
      return new Builder() {
        @Override
        public FileEditor build() {
          return provider.createEditor(project, file);
        }
      };
    }
  }
}

File: plugins/markdown/core/src/org/intellij/plugins/markdown/ui/split/SplitTextEditorProvider.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.intellij.plugins.markdown.ui.split

import com.intellij.openapi.editor.Document
import com.intellij.openapi.fileEditor.*
import com.intellij.openapi.progress.runBlockingCancellable
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.platform.ide.progress.runWithModalProgressBlocking
import com.intellij.util.application
import org.jdom.Element
import org.jetbrains.annotations.ApiStatus

private const val FIRST_EDITOR = "first_editor"
private const val SECOND_EDITOR = "second_editor"
private const val SPLIT_LAYOUT = "split_layout"

@ApiStatus.Internal
abstract class SplitTextEditorProvider(
  private val firstProvider: FileEditorProvider,
  private val secondProvider: FileEditorProvider
): AsyncFileEditorProvider, DumbAware {
  private val editorTypeId = createSplitEditorProviderTypeId(firstProvider.editorTypeId, secondProvider.editorTypeId)

  override fun accept(project: Project, file: VirtualFile): Boolean {
    return firstProvider.accept(project, file) && secondProvider.accept(project, file)
  }

  override fun acceptRequiresReadAction(): Boolean {
    return firstProvider.acceptRequiresReadAction() || secondProvider.acceptRequiresReadAction()
  }

  override fun createEditor(project: Project, file: VirtualFile): FileEditor {
    return createEditorAsync(project = project, file = file).build()
  }

  override fun getEditorTypeId(): String {
    return editorTypeId
  }

  override fun createEditorAsync(project: Project, file: VirtualFile): AsyncFileEditorProvider.Builder {
    val firstBuilder = createEditorBuilder(provider = firstProvider, project = project, file = file, document = null)
    val secondBuilder = createEditorBuilder(provider = secondProvider, project = project, file = file, document = null)
    return object : AsyncFileEditorProvider.Builder() {
      override fun build(): FileEditor {
        return createSplitEditor(firstEditor = firstBuilder.build(), secondEditor = secondBuilder.build())
      }
    }
  }

  override suspend fun createEditorBuilder(project: Project, file: VirtualFile, document: Document?): AsyncFileEditorProvider.Builder {
    val firstBuilder = createEditorBuilderAsync(provider = firstProvider, project = project, file = file, document = null)
    val secondBuilder = createEditorBuilderAsync(provider = secondProvider, project = project, file = file, document = null)
    return object: AsyncFileEditorProvider.Builder() {
      override fun build(): FileEditor {
        return createSplitEditor(firstEditor = firstBuilder.build(), secondEditor = secondBuilder.build())
      }
    }
  }

  protected fun readFirstProviderState(sourceElement: Element, project: Project, file: VirtualFile): FileEditorState? {
    val child = sourceElement.getChild(FIRST_EDITOR) ?: return null
    return firstProvider.readState(/* sourceElement = */ child, /* project = */ project, /* file = */ file)
  }

  protected fun readSecondProviderState(sourceElement: Element, project: Project, file: VirtualFile): FileEditorState? {
    val child = sourceElement.getChild(SECOND_EDITOR) ?: return null
    return secondProvider.readState(/* sourceElement = */ child, /* project = */ project, /* file = */ file)
  }

  protected fun readSplitLayoutState(sourceElement: Element, project: Project, file: VirtualFile): String? {
    return sourceElement.getAttribute(SPLIT_LAYOUT)?.value
  }

  override fun readState(sourceElement: Element, project: Project, file: VirtualFile): FileEditorState {
    val firstState = readFirstProviderState(sourceElement, project, file)
    val secondState = readSecondProviderState(sourceElement, project, file)
    val layoutName = readSplitLayoutState(sourceElement, project, file)
    return SplitFileEditor.MyFileEditorState(/* splitLayout = */ layoutName, /* firstState = */ firstState, /* secondState = */ secondState)
  }

  protected fun writeFirstProviderState(state: FileEditorState?, project: Project, targetElement: Element) {
    val child = Element(FIRST_EDITOR)
    if (state != null) {
      firstProvider.writeState(state, project, child)
      targetElement.addContent(child)
    }
  }

  protected fun writeSecondProviderState(state: FileEditorState?, project: Project, targetElement: Element) {
    val child = Element(SECOND_EDITOR)
    if (state != null) {
      secondProvider.writeState(state, project, child)
      targetElement.addContent(child)
    }
  }

  protected fun writeSplitLayoutState(splitLayout: String?, targetElement: Element) {
    if (splitLayout != null) {
      targetElement.setAttribute(SPLIT_LAYOUT, splitLayout)
    }
  }

  override fun writeState(state: FileEditorState, project: Project, targetElement: Element) {
    if (state !is SplitFileEditor.MyFileEditorState) {
      return
    }
    writeFirstProviderState(state = state.firstState, project = project, targetElement = targetElement)
    writeSecondProviderState(state = state.secondState, project = project, targetElement = targetElement)
    writeSplitLayoutState(splitLayout = state.splitLayout, targetElement = targetElement)
  }

  protected abstract fun createSplitEditor(firstEditor: FileEditor, secondEditor: FileEditor): FileEditor

  override fun getPolicy(): FileEditorPolicy {
    return FileEditorPolicy.HIDE_DEFAULT_EDITOR
  }
}

private fun createEditorBuilder(
  provider: FileEditorProvider,
  project: Project,
  file: VirtualFile,
  document: Document?
): AsyncFileEditorProvider.Builder {
  if (provider is AsyncFileEditorProvider) {
    if (application.isDispatchThread) {
      return runWithModalProgressBlocking(project, title = "Creating Markdown Editor") {
        provider.createEditorBuilder(project = project, file = file, document = document)
      }
    }
    return runBlockingCancellable {
      provider.createEditorBuilder(project = project, file = file, document = document)
    }
  }
  return object: AsyncFileEditorProvider.Builder() {
    override fun build(): FileEditor {
      return provider.createEditor(project, file)
    }
  }
}

private suspend fun createEditorBuilderAsync(
  provider: FileEditorProvider,
  project: Project,
  file: VirtualFile,
  document: Document?,
): AsyncFileEditorProvider.Builder {
  if (provider is AsyncFileEditorProvider) {
    return provider.createEditorBuilder(project = project, file = file, document = document)
  }
  return object: AsyncFileEditorProvider.Builder() {
    override fun build(): FileEditor {
      return provider.createEditor(project, file)
    }
  }
}

@ApiStatus.Internal
fun createSplitEditorProviderTypeId(first: String, second: String): String {
  return "split-provider[$first;$second]"
}
